src/System.Net.Http.Json/tests/FunctionalTests/TranscodingReadStreamTests.cs

```diff
@@ -0,0 +1,273 @@
+        {
+            new string('a', TranscodingReadStream.MaxCharBufferSize - 1) + "☀",
+            new string('a', TranscodingReadStream.MaxCharBufferSize - 2) + "☀",
+            new string('a', TranscodingReadStream.MaxCharBufferSize) + "☀",
+        };
+
+        [Theory]
+        [MemberData(nameof(ReadAsync_WithOverflowBuffer_AtBoundariesData))]
+        public Task ReadAsync_WithOverflowBuffer_WithBufferSize1(string input) => ReadAsync_WithOverflowBufferAtCharBufferBoundaries(input, bufferSize: 1);
+
+        [Theory]
+        [MemberData(nameof(ReadAsync_WithOverflowBuffer_AtBoundariesData))]
+        public Task ReadAsync_WithOverflowBuffer_WithBufferSize2(string input) => ReadAsync_WithOverflowBufferAtCharBufferBoundaries(input, bufferSize: 1);
+
+        private static async Task<TranscodingReadStream> ReadAsync_WithOverflowBufferAtCharBufferBoundaries(string input, int bufferSize)
+        {
+            // Arrange
+            // Test ensures that the overflow buffer works correctly
+            var encoding = Encoding.Unicode;
+            var stream = new TranscodingReadStream(new MemoryStream(encoding.GetBytes(input)), encoding);
+            var bytes = new byte[1];
+            var expected = Encoding.UTF8.GetBytes(input);
+
+            // Act
+            int read;
+            var buffer = new byte[bufferSize];
+            var actual = new List<byte>();
+
+            while ((read = await stream.ReadAsync(buffer, 0, bufferSize)) != 0)
+            {
+                actual.AddRange(buffer);
+            }
+
+            Assert.Equal(expected, actual);
+            return stream;
+        }
+
+        public static TheoryData ReadAsyncInputLatin =>
+            GetLatinTextInput(TranscodingReadStream.MaxCharBufferSize, TranscodingReadStream.MaxByteBufferSize);
+
+        public static TheoryData ReadAsyncInputUnicode =>
+            GetUnicodeText(TranscodingReadStream.MaxCharBufferSize);
+
+        internal static TheoryData GetLatinTextInput(int maxCharBufferSize, int maxByteBufferSize)
+        {
+            return new TheoryData<string>
+            {
+                "Hello world",
+                string.Join(string.Empty, Enumerable.Repeat("AB", 9000)),
+                new string('A', count: maxByteBufferSize),
+                new string('A', count: maxCharBufferSize),
+                new string('A', count: maxByteBufferSize + 1),
+                new string('A', count: maxCharBufferSize + 1),
+            };
+        }
+
+        internal static TheoryData GetUnicodeText(int maxCharBufferSize)
+        {
+            return new TheoryData<string>
+            {
+                new string('Æ', count: 7),
+                new string('A', count: maxCharBufferSize - 1) + 'Æ',
+                "AbĀāĂăĄąĆŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſAbc",
+               "Abcஐஒஓஔகஙசஜஞடணதநனபமயரறலளழவஷஸஹ",
+               "☀☁☂☃☄★☆☇☈☉☊☋☌☍☎☏☐☑☒☓☚☛☜☝☞☟☠☡☢☣☤☥☦☧☨☩☪☫☬☭☮☯☰☱☲☳☴☵☶☷☸",
```

Comment:
Please convert these to `\uxxxx` form so that the file can be saved as ASCII. We required that in this repo as historically files have gotten garbled otherwise. Hopfeully there's an easy way to convert them..

Reply:
Done in master and in blazor branch. 
Thanks.

Reply:
Thanks, guaranteed you would have gotten unit tests failing on some random day otherwise.